plugin/docker/volume.go

```diff
@@ -202,56 +206,67 @@ func (v *volume) VolumeRead(ctx context.Context, path string) ([]byte, error) {
 		activity.Record(ctx, "Closed file %v on %v: %v", mountpoint+path, cid, rdr.Close())
 	}()
 
+	// Read one file from the archive.
 	tarReader := tar.NewReader(rdr)
-	// Only expect one file.
-	if _, err := tarReader.Next(); err != nil {
+	if _, err = tarReader.Next(); err != nil {
 		return nil, err
 	}
-	bits, err := ioutil.ReadAll(tarReader)
-	if err != nil {
-		return nil, err
-	}
-	return bits, nil
+	return ioutil.ReadAll(tarReader)
 }
 
 func (v *volume) VolumeStream(ctx context.Context, path string) (io.ReadCloser, error) {
 	// Create a container that mounts a volume and tails a file. Run it and capture the output.
-	cid, err := v.createContainer(ctx, []string{"tail", "-f", mountpoint + path})
+	cid, cleanup, err := v.createContainer(ctx, []string{"tail", "-f", mountpoint + path})
 	if err != nil {
 		return nil, err
 	}
 
-	// Manually use this in case of errors. On success, the returned Closer will need to call instead.
-	delete := func(ct context.Context) {
-		err := v.client.ContainerRemove(ct, cid, types.ContainerRemoveOptions{})
-		activity.Record(ctx, "Deleted container %v: %v", cid, err)
-	}
-
-	activity.Record(ctx, "Starting container %v", cid)
-	if err := v.client.ContainerStart(ctx, cid, types.ContainerStartOptions{}); err != nil {
-		activity.Record(ctx, "Error starting container %v: %v", cid, err)
-		delete(context.Background())
-		return nil, err
-	}
-
-	// Manually use this in case of errors. On success, the returned Closer will need to call instead.
-	killAndDelete := func() {
-		ct := context.Background()
-		err := v.client.ContainerKill(ct, cid, "SIGKILL")
-		activity.Record(ctx, "Stopped temporary container %v: %v", cid, err)
-		delete(ct)
-	}
-
 	opts := types.ContainerLogsOptions{ShowStdout: true, Follow: true, Tail: "10"}
 	activity.Record(ctx, "Streaming log for %v", cid)
 	output, err := v.client.ContainerLogs(ctx, cid, opts)
 	if err != nil {
-		killAndDelete()
+		cleanup()
 		return nil, err
 	}
 
 	// Wrap the log output in a ReadCloser that stops and kills the container on Close.
-	return plugin.CleanupReader{ReadCloser: output, Cleanup: killAndDelete}, nil
+	return plugin.CleanupReader{ReadCloser: output, Cleanup: cleanup}, nil
+}
+
+func (v *volume) VolumeWrite(ctx context.Context, path string, b []byte, mode os.FileMode) error {
+	// Create a container that mounts a volume and waits. Use it to upload a file.
+	cid, cleanup, err := v.createContainer(ctx, []string{"sleep", "60"})
+	if err != nil {
+		return err
+	}
+	defer cleanup()
+
+	// Create a tar of the file contents and upload it.
```

Comment:
Why are we tar'ing and copying? Some clarification would help. Seems weird that this is different from the Kubernetes PVC implementation.

Reply:
Other than the clarification, PR looks good.

Reply:
Added. I suppose I could've done it with execs like everything else. That would avoid needing to pass in the permissions.

Reply:
How hard would it be to do that? The inconsistency between this and Kube PVC is kind of weird.

Reply:
I'll explore cleaning this up similar to Kubernetes.

I didn't add the same logic to identify an existing container because
- Docker scheduling is a lot faster, so it's not much overhead compared to starting a new process
- Docker doesn't have anything like ReadWriteOnce, so we can always get read/write access